torch/csrc/quantized/quantized_backward.cpp

```diff
@@ -0,0 +1,121 @@
+  // Backprop adjoint through layer.
+  at::Tensor matmul_backward(const at::Tensor& grad) {
+    auto& ctx = at::globalContext();
+    at::Tensor weight;
+    c10::optional<at::Tensor> bias;
+    std::tie(weight, bias) = self_->unpack();
+    weight = at::permute(weight, {1, 0});
+    int gradN = grad.sizes().vec().size();
+    std::vector<int64_t> gradArgs;
+    for (int i = 0; i < gradN; i++) {
+      gradArgs.push_back(i);
+    }
+    c10::intrusive_ptr<LinearPackedParamsBase> prepackedWeightTransposed;
+    if (ctx.qEngine() == at::QEngine::QNNPACK) {
+      TORCH_CHECK(
+          weight.scalar_type() == at::kQInt8,
+          "QNNPACK only supports INT8 bit width currently. Got ",
+          c10::toString(weight.scalar_type()));
+      prepackedWeightTransposed =
+          PackedLinearWeightsQnnp::prepack(std::move(weight), nullopt);
+      return prepackedWeightTransposed->apply_dynamic(grad);
+    }
+    TORCH_CHECK(
+        false,
+        "Didn't find engine for operation quantized::linear_prepack ",
+        toString(ctx.qEngine()));
+  }
+
+  // This is invoked by the PyTorch backward() process.
+  variable_list apply(variable_list&& grads /*adjoint*/) override {
+    std::lock_guard<std::mutex> lock(mutex_);
+    IndexRangeGenerator gen;
+    auto input_ix = gen.range(1);
+    variable_list grad_inputs(gen.size());
+    bool any_grad_defined = any_variable_defined(grads);
+    const auto& grad = grads[0]; // adjoint
+    if (should_compute_output({input_ix})) {
+      auto grad_result =
+          any_grad_defined ? matmul_backward(grad) : at::Tensor();
+      copy_range(grad_inputs, input_ix, grad_result);
+    }
+    return grad_inputs;
+  }
+  std::string name() const override {
+    return "PackedLinearWeightDynamicBackward";
+  }
+  void release_variables() override {
+    std::lock_guard<std::mutex> lock(mutex_);
+    // resetting weight is necessary?
+  }
+  // SavedVariable self_; // TODO: Any ref counting needed?
+  c10::intrusive_ptr<LinearPackedParamsBase> self_;
+  std::vector<int64_t> input_sizes_;
+};
+
+} // namespace
+
+at::Tensor packed_linear_weight_grad(
+    at::Tensor input,
+    const c10::intrusive_ptr<LinearPackedParamsBase>& packed_weight,
+    bool reduce_range) {
+  at::AutoDispatchBelowADInplaceOrView guard;
+  auto _any_requires_grad = compute_requires_grad(input);
+  std::shared_ptr<PackedLinearWeightDynamicBackward> grad_fn;
+
+  if (_any_requires_grad) {
+    grad_fn = std::shared_ptr<PackedLinearWeightDynamicBackward>(
+        new PackedLinearWeightDynamicBackward(), deleteNode);
+    grad_fn->self_ = packed_weight;
+    grad_fn->input_sizes_ = input.sizes().vec();
+    grad_fn->set_next_edges(collect_next_edges(input));
+  }
+  static auto op = at::Dispatcher::singleton()
+                       .findSchemaOrThrow("quantized::linear_dynamic", "")
+                       .typed<decltype(packed_linear_weight_grad)>();
+  auto output = op.call(input, packed_weight, reduce_range);
```

Comment:
This should be a redispatch.

Reply:
I think this was last time I was suggested to use this way?, I'm unsure how to locate redispatch function for the scheme (can't find anywhere honestly)

Reply:
Yes in that link, it calls `op.redispatch` not `op.call`